File: go/point.go
package main

import "fmt"

const maxLen = 15

type direction struct {
	x, y int
}

var fourDirections = []direction{{-1, 0}, {-1, -1}, {0, -1}, {1, -1}}

var eightDirections = []direction{{-1, 0}, {-1, -1}, {0, -1}, {1, -1}, {1, 0}, {1, 1}, {0, 1}, {-1, 1}}

type point struct {
	x, y int
}

func (p point) move(dir direction, length int) point {
	if length == 0 {
		return p
	}
	return point{p.x + dir.x*length, p.y + dir.y*length}
}

func (p point) checkRange() bool {
	return p.x < maxLen && p.x >= 0 && p.y < maxLen && p.y >= 0
}

func (p point) nearMidThan(p2 point) bool {
	return max(abs(p.x-maxLen/2), abs(p.y-maxLen/2)) < max(abs(p2.x-maxLen/2), abs(p2.y-maxLen/2))
}

func (p point) hash() int {
	return p.y*maxLen + p.x
}

func (p point) String() string {
	return fmt.Sprintf("(%d,%d)", p.x, p.y)
}

func max(x ...int) int {
	m := x[0]
	for i := 1; i < len(x); i++ {
		if x[i] > m {
			m = x[i]
		}
	}
	return m
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: java/src/com/github/reimu/gobang/Point.java
package com.github.reimu.gobang;

public final class Point {
	public final int x;
	public final int y;
	public Point(int x, int y) {
		this.x = x;
		this.y = y;
	}
	Point move(Direction dir, int len) {
		if (len == 0) return new Point(x, y);
		int x = this.x + dir.x * len;
		int y = this.y + dir.y * len;
		if (!checkRange(x, y)) return null;
		return new Point(x, y);
	}
	public boolean nearMidThan(Point p) {
		return Math.max(Math.abs(x - Constant.MAX_LEN / 2), Math.abs(y - Constant.MAX_LEN / 2))
			< Math.max(Math.abs(p.x - Constant.MAX_LEN / 2), Math.abs(p.y - Constant.MAX_LEN / 2));
	}
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof Point))
			return false;
		return ((Point) obj).x == x && ((Point) obj).y == y;
	}
	@Override
	public int hashCode() {
		return y * Constant.MAX_LEN + x;
	}
	@Override
	public String toString() {
		return toString(x, y);
	}
	private static String toString(int x, int y) {
		return "(" + x + ',' + y + ')';
	}
	public static boolean checkRange(int x, int y) {
		return x < Constant.MAX_LEN && x >= 0 && y < Constant.MAX_LEN && y >= 0;
	}
}
